### ISSUE TITLE
Add documentation & example on writing tests for Hamilton code

### ISSUE THREAD

**Issue by skrawcz** [DAGWorks-Inc/hamilton#1044]
**Is your feature request related to a problem? Please describe.**
We don't have an example and documentation around testing hamilton code. It's more straightforward than people think, e.g. some think that with decorators testing becomes harder.

**Describe the solution you'd like**
1. A section in the User Guide on writing tests.
2. This should cover unit testing functions without decorators, and functions with decorators.
3. This should cover integration testing parts of a DAG - using overrides to show how to run various parts of the graph.
4. This should cover how test custom materializers -- basically just show an example of how we test the in built ones. E.g. create a resource folder with python code, or use create_adhoc_module from adhoc_utils.

**Describe alternatives you've considered**
N/A

**Additional context**
You can largely see how the Hamilton tests work for an idea for content/material.

**Comment by audacious-fk** [comment]
Hi!
I would love to work on this issue!
is it open?

**Comment by skrawcz** [comment]
> Hi! I would love to work on this issue! is it open?

yes it is. Do you want to start with https://github.com/DAGWorks-Inc/hamilton/issues/212 first? and then come back to this?

**Comment by audacious-fk** [comment]
Yeah, sure!
